packages/@aws-cdk/aws-secretsmanager/lib/secret.ts

```diff
@@ -237,6 +237,20 @@ abstract class SecretBase extends Resource implements ISecret {
   protected get arnForPolicies() { return this.secretArn; }
 }
 
+/** Options for `Secret.fromSecretName` */
+export interface SecretFromSecretNameOptions {
+  /**
+   * Whether the secret imported by name will respond with a "partial" ARN without the Secrets Manager suffix.
+   * If this is set, `secretArn` will return a composed ARN without the Secrets Manager suffix.
+   * If not set, 'secretArn' will return the `secretName`.
```

Comment:
Don't like this quirk. It's so non-trivial and easy to miss. 
`arn` should return ARN, not name.

How about instead introduce three things in `ISecret` - `secretName`, `secretPartialArn` and `secretArn`?
And then have the combination of import APIs as needed. 

However, if a customer imports a secret with only the name or the partial ARN and their integration (i.e., ECS) requests a fullArn, the API will fail.

Reply:
> However, if a customer imports a secret with only the name or the partial ARN and their integration (i.e., ECS) requests a fullArn, the API will fail.

This means customers who currently import the secret by name and make some reference to `secretArn` somewhere will go from having a working synth/deploy to an error.

The only way to make the above work -- I think -- would be to deprecate `fromSecretName` and introduce a new `fromSecretNameV2` or `fromSecretNameWithProperPartialArnHandling` (or `fromSecretNameButNotDeprecated`) or something. Any suggestions? 

--- 

This also doesn't quite solve how the integration knows what forms are available. Requesting the `secretArn` and catching an error doesn't fit. The only sane thing I can think of is that integrations must always just use the "weakest" form they can accept. For example, CodeBuild would always just use the `secretName`, even if the full ARN was available.

This has some implications, especially given that today users can provide either a partial or full ARN to `fromSecretArn`, and the code that determines the `secretName` from the ARN can't deterministically know which was provided. An ARN of `arn:aws:secretsmanager:eu-west-1:111111111111:secret:github-tokens` will be treated like a full ARN with a SecretsManager suffix of 'tokens', rather than a partial ARN.

```ts
const secret = secretsmanager.Secret.fromSecretArn(this, 'secret', 'arn:aws:secretsmanager:eu-west-1:111111111111:secret:github-tokens');
secret.secretName; // 'github'
```

There are likely scenarios where an integrating service will break due to the above, but be fine with the `secretArn` (or even `partialArn`).

In lieu of error checking, I suppose we could introduce new fields to track what is available (e.g., `hasFullArn`), but that feels gross.

What do you think?

Reply:
I think it's worth it to look from the perspective of the consumers of the SecretsManager library. How should using it look like?

For example, we have code in CodeBuild that uses `ISecret`: https://github.com/aws/aws-cdk/blob/0537598a065c74c713d1ec58c96f09f0d66b12ca/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L966-L966

Is there any way we can sensibly make a decision in that CodeBuild Project code to use either `secretName` or `secretArn`, considering CodeBuild is one of the services that supports both?

Reply:
@nija-at / @skinny85 - How much do you both hate something like this?

```ts
  /**
   * Returns an identifier for the secret that can be used to reference the secret in other constructs.
   * For owned secrets, this will always be the secret ARN. For imported secrets,this identifier may be either
   * the full secret ARN (with SecretsManager-suffix),
   * partial secret ARN (without the suffix),
   * or just the secret name, depending on what values are available.
   * Returns the "strongest" identifier available (full > partial > secret name), unless `acceptPartialArns` is false.
   */
  public secretId(acceptPartialArns: boolean = true): string {
    return (!this._isPartialArn || acceptPartialArns) ? this.secretArn : this.secretName;
  }
```

`_isPartialArn` would only be true for secrets imported by name; a new `fromPartialSecretArn` could also be created as well to set the flag but otherwise behave like `fromSecretArn`.

This moves the responsibility from the end-user to construct library author.

The CodeBuild code would be updated to read like this:

```ts
credential: this.buildImage.secretsManagerCredentials.secretId(false), 
```

And the ECS code could be updated as well:

https://github.com/aws/aws-cdk/blob/a4a41b5e006110304b51ee55c34e91cc3f129281/packages/@aws-cdk/aws-ecs/lib/container-definition.ts#L39

```ts
arn: field ? `${secret.secretId()}:${field}::` : secret.secretId(),
```

--- 

The above could be done in conjunction with -- or independently from -- creating a `fromSecretNameV2` (better name suggestions welcome).

Reply:
Oh man. This is such a mess.

I actually think the simplest solution would be to have `secretName` return `string | undefined`. I actually thought that was already the case, because of #10914. So then, in CodeBuild, you could do:

```ts
credential: this.buildImage.secretsManagerCredentials.secretName ?? this.buildImage.secretsManagerCredentials.secretArn,
```

 (I assume a new Secret without a `secretName` set in `SecretProps` would return an undefined `secret.secretName`)

While on ECS you would always use `secretArn`.

Is it too late to go this route?

Reply:
> I actually think the simplest solution would be to have secretName return string | undefined. I actually thought that was already the case, because of #10914.

:cry: Well, it *is*, but only for owned `Secrets`. Imported secrets always set the `secretName`. This is frankly just an oversight on my part; owned `Secrets` were intended to have `secretName` set as well; I think I mistakenly believed that the below line would treat `this.physicalName` as a `Token` and set the name after creation. Whoops.

https://github.com/aws/aws-cdk/blob/a4a41b5e006110304b51ee55c34e91cc3f129281/packages/@aws-cdk/aws-secretsmanager/lib/secret.ts#L354

> Is it too late to go this route?

With the existing `fromSecretName`, yes. With a new `fromSecretNameV2`, no, but I question whether it's the right route. It does make that above usage possible, but at the cost of not having a `secretName` in cases where we definitely could.

Reply:
I wouldn't mind if @eladb weighed in here with some thoughts...

Reply:
Definitely feels wrong to put the burden on the user. This is precisely stuff we would want to completely encapsulate within the library. But I am not sure I completely understand all the implications.

@njlynch can you please summarize what would be the ideal API, if we had designed it from the ground up? Then, can you describe which aspects of this API break the existing behavior.

Reply:
**Background:**

Secrets are given a name on creation, either by users or generated by CloudFormation. The name is then suffixed with a hyphen plus a six-character string to form the "full" name (e.g., `MySecret` becomes `MySecret-abc123`). The ARN is formatted as `arn:aws:secretsmanager:$REGION:$ACCOUNT:secret:$SECRET_NAME`. Depending on the scenario (which API or CloudFormation integration you're using), the secret name (with and without the suffix) or ARN (with and without the suffix) can be used. Additional complexity comes from the fact that secret names can contain hyphens themselves, so 'my-secret-abc123' could be a "partial" secret name or a "full" secret name, and there's no deterministic way to know. Secrets Manager [discourages](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DescribeSecret.html#SecretsManager-DescribeSecret-request-SecretId) creating secrets of this format, but doesn't prevent it.

Given a secret name (without a 6-character suffix), a partial ARN may be created. Given an ARN, a partial ARN and secret name can be derived; however, this runs into issues when the secret name itself contains a 6-character suffix (which is valid).

**Desired state:**

Users want to be able to import existing secrets by both (partial|full) ARN and (partial|full) name. Importing by partial name is often the most convenient for multi-environment setups, as the partial name is the only identifier that may be the same across accounts and regions. Ideally, a user imports the secret given a (partial|full) (name|ARN) and then is able to use it anywhere an `ISecret` is accepted, without worrying about the underlying details.

Various constructs have different rules for which secret identifier they can accept. CodeBuild will not accept a partial ARN but can accept a secret name. ECS will only accept an ARN, either partial or full. The onus of determining which identifier(s) are available and which to use should fall on the construct library, not the user. Generally, the construct library should prefer to use the "strongest" (full ARN > partial ARN > secret name) identifier available, as it reduces ambiguity.

Ideal API:

```ts
interface ISecret {
  readonly encryptionKey?: kms.IKey;
  readonly secretArn: string;
  readonly secretName: string;
  readonly secretValue: SecretValue;
  // True iff the `secretArn` does not contain the SecretsManager-provided suffix.
  readonly secretArnIsPartial: boolean;
}

// Open to slightly different representations, see https://github.com/aws/aws-cdk/pull/10410#discussion_r490465209
interface SecretAttributes {
  readonly secretArn?: string;
  readonly encryptionKey?: kms.IKey;
  readonly secretName?: string;
}

// Imports the secret by name, sets `secretArnIsPartial = true`.
public static fromSecretName(scope: Construct, id: string, secretName: string)
// Imports the secret by ARN, automatically sets the secret name. `secretArnIsPartial = false`
public static fromSecretArn(scope: Construct, id: string, secretArn: string)
// Imports the secret by ARN or name.
// If secretName and secretArn are both provided, explicitly sets secretName,
// and sets `secretArnIsPartial` based on the comparison of the ARN and secretName.
// Use case is for handling the worst-case secret names with 6-character suffixes.
public static fromSecretAttributes(scope: Construct, id: string, attrs: SecretAttributes)

constructor(scope: Construct, id: string, props: SecretProps) {
 // secretArnIsPartial = false;
}
```

Consumers of the `ISecret` use the `secretArnIsPartial` flag to make a determination on how to handle the secret. For example, CodeBuild's logic would be:

```ts
credential: secret.secretArnIsPartial ? secret.secretName : secret.secretArn;
```

ECS would simply continue using `secretArn` as today.

**Why can't we do this today?:**

Today, `fromSecretName` sets `secretArn === secretName`. This breaks several existing integrations, but can't be unset because it also *works* for some integrations.

The easiest way around that is deprecating `fromSecretName` and introducing `fromSecretNameV2` which conforms to the above behavior.

---

Thoughts? Counter proposals?

Reply:
(discussed offline and here's where we landed)

The API is going to look like this -

```ts
export interface ISecret {
  readonly secretName: string;
  readonly secretPartialArn: string;

  // fromSecretName() and fromSecretPartialArn() throws an exception
  readonly secretFullArn: string;

  // This will return false only when a secret is imported using its name or partial ARN
  public trySecretFullArn(): boolean;

  // returns `trySecretFullArn() ?? secretPartialArn`;
  // so that 15 uses of `secretArn` across the CDK still works
  // and according to @njlynch either works just fine in most cases.
  readonly secretArn: string; 
}
```

Following this the `fromSecretName()` API will be deprecated and another `fromSecretNameV2()` introduced that does the correct thing. Also introduce, `fromSecretPartialArn()` and `fromSecretFullArn()`.

Reply:
I am [not yet] aware of any use case for `secretPartialArn`, and I don't believe creating it for owned secrets is possible (there's no equivalent of `lastIndexOf` as a CloudFormation `Fn`). So that means either having `secretPartialArn` throw for owned secrets, having a `trySecretPartialArn`, or just removing it entirely. Of the three options, the latter seems cleanest.

So this is @nija-at's (modified) `ISecret`.
```ts
export interface ISecret {
  readonly encryptionKey?: kms.IKey;
  readonly secretName: string;
  readonly secretValue: SecretValue;
  // Behaves exactly as today. Will return a partial ARN from `fromSecretNameV2`.
  readonly secretArn: string; 
  // throws when used via fromSecretName() and fromSecretNameV2(),
  // returns secretArn otherwise.
  readonly secretFullArn: string;
  // return undefined when used via fromSecretName() and fromSecretNameV2(),
  // returns secretArn otherwise.
  readonly trySecretFullArn?: string;
}
```

And this is my proposal for comparison `ISecret`:
```ts
export interface ISecret {
  readonly encryptionKey?: kms.IKey;
  readonly secretName: string;
  readonly secretValue: SecretValue;
  // Behaves exactly as today. Will return a partial ARN from `fromSecretNameV2`.
  readonly secretArn: string;
  // True when used via `fromSecretNameV2`.
  readonly secretArnIsPartial: boolean;
}
```